p1-staking-pool-dyn/src/lib.rs

```diff
@@ -269,19 +271,30 @@ impl Contract {
             PromiseResult::NotReady => unreachable!(),
 
             PromiseResult::Successful(_) => {
-                let mut vault = self.vaults.get(&user).expect(ERR10_NO_ACCOUNT);
-                vault.rewards = vault.rewards.saturating_sub(amount.0);
-                self.vaults.insert(&user, &vault);
-                self.total_rewards += amount.0;
+                self.total_rewards += amount_cheddar.0;
                 if close {
                     self.vaults.remove(&user);
-                    env::log(b"account closed");
-                    return;
+                    env_log!(
+                        "Account {} closed, farmed {} CHEDDAR",
+                        &user,
+                        amount_cheddar.0
+                    );
+                } else {
+                    env_log!("Farmed CHEDDAR: {}", amount_cheddar.0);
+                    let mut vault = self.vaults.get(&user).expect(ERR10_NO_ACCOUNT);
+                    vault.rewards = vault.rewards.saturating_sub(amount_cheddar.0);
+                    self.vaults.insert(&user, &vault);
+                }
+                if amount_near.0 != 0 {
+                    Promise::new(user).transfer(amount_near.0);
                 }
             }
 
             PromiseResult::Failed => {
```

Comment:
PromiseResult::Failed => { needs to panic! so the FE & the user know what happened
that's why you cannot alter state on PromiseResult::Failed

Reply:
1. FE won't know what happened, because it's called before (this is the last callback)
2. user knows what happened because I added a log - panic doesn't help here, it won't rollback any changes from the previous "receipts".

Reply:
the user and the FE dont see logs. You need to panic! to properly inform of a failed txn